tools/idf_size.py

```diff
@@ -109,37 +129,34 @@ def load_sections(map_file):
             sections[section["name"]] = section
             continue
 
-        # source file line, ie
-        # 0x0000000040080400       0xa4 /home/gus/esp/32/idf/examples/get-started/hello_world/build/esp32/libesp32.a(cpu_start.o)
-        RE_SOURCE_LINE = r"\s*(?P<sym_name>\S*).* +0x(?P<address>[\da-f]+) +0x(?P<size>[\da-f]+) (?P<archive>.+\.a)\((?P<object_file>.+\.ob?j?)\)"
-
-        m = re.match(RE_SOURCE_LINE, line, re.M)
-        if not m:
-            # cmake build system links some object files directly, not part of any archive
-            RE_SOURCE_LINE = r"\s*(?P<sym_name>\S*).* +0x(?P<address>[\da-f]+) +0x(?P<size>[\da-f]+) (?P<object_file>.+\.ob?j?)"
-            m = re.match(RE_SOURCE_LINE, line)
-        if section is not None and m is not None:  # input source file details=ma,e
-            sym_name = m.group("sym_name") if len(m.group("sym_name")) > 0 else sym_backup
-            try:
-                archive = m.group("archive")
-            except IndexError:
-                archive = "(exe)"
+        if section is not None:
+            m = RE_SYMBOL_ONLY_LINE.match(line)
+            if m is not None:
+                # In some cases the section name appears on the previous line, back it up in here
+                sym_backup = m.group("sym_name")
+                continue
 
-            source = {
-                "size": int(m.group("size"), 16),
-                "address": int(m.group("address"), 16),
-                "archive": os.path.basename(archive),
-                "object_file": os.path.basename(m.group("object_file")),
-                "sym_name": sym_name,
-            }
-            source["file"] = "%s:%s" % (source["archive"], source["object_file"])
-            section["sources"] += [source]
-
-        # In some cases the section name appears on the previous line, back it up in here
-        RE_SYMBOL_ONLY_LINE = r"^ (?P<sym_name>\S*)$"
-        m = re.match(RE_SYMBOL_ONLY_LINE, line)
-        if section is not None and m is not None:
-            sym_backup = m.group("sym_name")
+            if not RE_PRE_FILTER.match(line):
+                # line does not match our quick check, so skip to next line
```

Comment:
Just out of total curiosity, what's the performance difference from leaving out the pre-filter step?

Reply:
> tl;dr  without the pre-filter it goes from 375 ms to 561 ms

So i started looking into optimizing it with various small things here and there before I narrowed down the issue was with that specific regex (`RE_SOURCE_FILE`). I ended up getting fairly good gains, but not in the ballpark of what I have now. It wasn't until I discovered one very small thing that gave very high gains: there was an extra un-needed wildcard after the symbol name.

```python                                                      
                                       |
                                       V
RE_SOURCE_LINE = r"\s*(?P<sym_name>\S*).* +0x(?P<address> ....
```

So by combining the cmake and source file regexes into 1 it cut the time down in almost half, but I just checked and with this small regex tweak and nothing else changing, even leaving the cmake as a separate regex (with the fix), brings it down to ~1.4 seconds. 

Of course by the time i discovered this I had already made a lot of the other small tweaks whose performance gains became a lot less significant.

I played around with it a bit, commenting out fixes, and wrote a bash script to run it 10 times and print the average run. The results are kinda interesting to see. The first row is the code as it is now, the following rows an `⨯` represents a fix/tweak that I undid for bench-marking. As you can see without the regex fix the pre-filter and early exit made a much larger difference than they do once you include the regex fix.

Pre-Filter | Early Exit | Fixed Regex | Avg Millis
-- | -- | -- | --
✓ | ✓ | ✓ | 375
⨯ | ✓ | ✓ | 561
✓ | ⨯ | ✓ | 505
⨯ | ⨯ | ✓ | 675
✓ | ✓ | ⨯ | 1291
⨯ | ✓ | ⨯ | 4672
✓ | ⨯ | ⨯ | 6023
⨯ | ⨯ | ⨯ | 9294

Reply:
Thanks for the very comprehensive writeup!